### ISSUE TITLE
Unable to search for`cassini` LDD attributes in ISS datasets

### ISSUE THREAD

**Issue by jordanpadams** [NASA-PDS/registry-sweepers#148]
### Checked for duplicates

Yes - I've already checked

### 🐛 Describe the bug

When I tried to search by `cassini:ISS_Specific_Attributes.cassini:image_number` via the API, I get no results, when I should get many.

### 🕵️ Expected behavior

I expected to be able to search by this field and get a result

### 📜 To Reproduce

https://pds.mcp.nasa.gov/api/search/1/products?q=(cassini:ISS_Specific_Attributes.cassini:image_number%20eq%20%221454725799%22) should return 1 result: https://pds-rings.seti.org/pds4/bundles/cassini_iss_saturn//data_raw/14547xxxxx/1454725799n.xml

Same thing in Kibana Discover, no go.

### 🖥 Environment Info

- Version of this software [e.g. vX.Y.Z]
- Operating System: [e.g. MacOSX with Docker Desktop vX.Y]
...

### 🦄 Related requirements

🦄 NASA-PDS/registry-api#539

### ⚙️ Engineering Details

I am concerned more broadly that attributes throughout the systems are randomly unsearchable because harvest is not or was not properly creating fields in the schema prior to loading them into the index. Not sure how we can scrub this, but a sweeper may be necessary to somehow scan and fix this all the time.

**Comment by alexdunnjpl** [comment]
Implemented in `index-mapping-repair` with the exception of two wrinkles:

resolution of missing property mapping typename TBD (@jordanpadams please weigh in on that)

@jordanpadams @tloubrieu-jpl the sweeper queries twice - once to generate the set of missing mappings, then again to generate/write the doc updates once the mappings have been ensured.  These two queries need to return consistent results, otherwise an old version of harvest could write new documents in the middle of a sweep which would get picked up in the second stage but not the first.

In that (theoretically-possible but shockingly-unlikely) event, those documents would erroneously be marked as fixed and excluded from future sweeps, and there only way to detect them would be to manually run the sweeper with the redundant-work filter disabled.  Pick an option, in increasing order of rigor:

1. The likelihood of someone running an obsolete version of harvest at exactly the wrong time is functionally zero - don't guard against it.

2. Instead of filtering to "documents which haven't been swept before", apply an additional constraint of "harvest time is earlier than sweeper execution start".

3. Use a point-in-time search.  

3 is the most-correct option, but may not be compatible with our dockerized registry, so I'd prefer to go with 2, or 1 if you're _absolutely_ sure no-one will run a pre-2023 version of harvest at just the wrong time.

**Comment by alexdunnjpl** [comment]
![image](https://github.com/user-attachments/assets/4fc51a16-6f2a-449c-a5b1-833db330f112)

**Comment by alexdunnjpl** [comment]
Log earliest/latest harvest timestamp for affected files, and a unique list of harvest versions.  Pull these from the docs themselves.

**Comment by alexdunnjpl** [comment]
Per @jordanpadams, 

>I think you can use the -dd indexes in the registry for tracking down these classes/attributes.

**Comment by alexdunnjpl** [comment]
status: implemented, in review
per @jordanpadams review/live-test postponed until next week, after the current site demos.

**Comment by alexdunnjpl** [comment]
status: testing against MCP rms in-progress.

The initial run got 40min and about halfway through, then AOSS throttled.  I doubt this is something we care to address if sweepers initialization is the only thing that hits whatever limits are imposed.

@jordanpadams @tloubrieu-jpl the query referred to in the OP now successfully hits on a single document.  If more hits are expected, assume that they'll start showing once my re-run of the sweepers against rms is complete - the throttling doesn't appear to persist for very long so that's not an impediment.

Once it checks out, want me to run it against all the other nodes, and include all the sweepers (not just the reindexer)?

**Comment by alexdunnjpl** [comment]
For rms, problems were detected for harvest version 3.8.1, and harvest timestamps 2022-06-28 through 2024-03-13.

Logs were long due to many documents not having harvest versions and throwing warnings, so I haven't sent them through - @jordanpadams let me know if you'd like me to strip those out and send them.

**Comment by alexdunnjpl** [comment]
Confirmed with `en` that a single run is sufficient to reindex all documents.   Currently running manually against all nodes, storing logs for later analysis

**Comment by alexdunnjpl** [comment]
Status: [running against large indices appears to overload those indices](https://github.com/NASA-PDS/registry/issues/329).  Need to figure out a way to consistently page through the documents.

Given the way it works:

- the workload can be chunked without issue if the first/second query can be guaranteed to return the same result-set (this means PIT, most-likely, or sorting by harvest timestamp if that's infeasible), since any product which has yet to be processed will eventually be updated/reindexed, and any product which is updated/reindexed is guaranteed to have had its appropriate mappings created already.

- if that is difficult or impossible, a naive approach which pages blindly could work iff the update generation step also checks that the mapping is present, not yielding an update if a missing mapping exists at update-creation-time

**Comment by alexdunnjpl** [comment]
[PIT search is only available as of OpenSearch 2.4](https://opensearch.org/blog/point-in-time/#:~:text=New%20for%20version%202.4%2C%20Point,going%20both%20forward%20and%20backward.), and while [AWS OpenSearch Service supports OpenSearch 2.15](https://aws.amazon.com/about-aws/whats-new/2024/09/amazon-opensearch-service-version-2-15/), [AWS Serverless Collections currently uses OpenSearch 2.0](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/serverless-overview.html)

> Serverless collections currently run OpenSearch version 2.0.x. As new versions are released, OpenSearch Serverless will automatically upgrade your collections to consume new features, bug fixes, and performance improvements.

so point-in-time search is not available to us at this point and a stopgap solution must be implemented.

**Comment by jordanpadams** [comment]
Status: continuing to test this more rigorously, and ran into some issues on production indexes. working on improving the algorithm to support this.

**Comment by alexdunnjpl** [comment]
Status: current run against ATM terminated as cluster is having to do a bunch of redundant work (indexing is slow, resulting in duplicated updates being written, resulting in more requests which probably affect indexing performance even if no-op)

Sweeper will be updated to pause until 95(?)% of the pending updates have processed and been reflected in the remaining hits count

ATM will require reindexing once @sjoshi-jpl is back [because reasons](https://github.com/NASA-PDS/planetary-data-cloud/issues/118)

**Comment by alexdunnjpl** [comment]
Status: flow management code is tested and operational. Completing ATM sweep, will review logs with @jordanpadams before running against other nodes.

**Comment by alexdunnjpl** [comment]
Status: flow management code imperfect, sometimes waits for a hits change which never comes.
Next step: implement a check which stops stalling if the same hits-count is returned n times in a row.  Stall functionality will need to be extracted to its own class at this point - it's getting sufficiently complicated.

**Comment by tloubrieu-jpl** [comment]
@alexdunnjpl is looking at how not to overload OpenSearch.

**Comment by alexdunnjpl** [comment]
Development complete, running successfully against ATM (which will need manual reindexing when @sjoshi-jpl returns), will run now against another larger node to fully-validate.

**Comment by tloubrieu-jpl** [comment]
Some bugs needed to be fixed while testing on IMG. That will be ready for merge shortly.

**Comment by tloubrieu-jpl** [comment]
Alex is running the sweepers in production.

**Comment by alexdunnjpl** [comment]
PSA is in-prog, should finish in the next day or two.

**Comment by alexdunnjpl** [comment]
@tloubrieu-jpl PR is ready for review/merge - will close that out now
